File: liqp-core/src/main/java/liqp/nodes/FilterNode.java
package liqp.nodes;

import com.google.common.collect.ImmutableList;
import java.util.List;
import liqp.node.LNode;
import liqp.filter.FilterInstance;
import liqp.filter.FilterParams;
import liqp.filter.LFilter;
import lombok.Builder;
import lombok.Getter;
import lombok.Singular;
import org.jetbrains.annotations.NotNull;

@Getter
public class FilterNode implements LNode {

  private final LFilter filter;
  private final List<LNode> params;

  @Builder
  public FilterNode(@NotNull LFilter filter, @Singular List<LNode> params) {
    this.filter = filter;
    this.params = ImmutableList.copyOf(params);
  }

  /**
   * Creates an instance of this filter for rendering.  This wrapper class allows for delayed resolution of
   * filter params, in case this filter is never invoked.
   */
  public FilterInstance getInstance() {
    return new FilterInstance(filter, new FilterParams(params));
  }

  @Override
  public List<LNode> children() {
    return ImmutableList.<LNode>builder()
          .addAll(params)
          .build();
  }

  @Override
  public Object render(LContext context) {
    throw new IllegalStateException("cannot render a filter");
  }
}

File: liqp-core/src/main/kotlin/liqp/nodes/FilterNode.kt
package liqp.nodes

import liqp.context.LContext
import liqp.filter.FilterInstance
import liqp.filter.FilterParams
import liqp.filter.LFilter
import liqp.filter.ResolvableFilterParams
import liqp.node.LNode

class FilterNode
constructor(val filter: LFilter, val params: List<LNode>) : LNode() {

  override val children: List<LNode> get() = params

  override fun render(context: LContext): Any = TODO()

  /**
   * Creates an instance of this filter for rendering.  This wrapper class allows for delayed resolution of
   * filter params, in case this filter is never invoked.
   */
  operator fun get(context: LContext): FilterInstance {
    return FilterInstance(filter, ResolvableFilterParams(context, params))
  }
}
